### PR TITLE
Fixes #193 chain QA test within travis build

### LINKED ISSUE #193: chain QA test within travis build
following https://github.com/geokrety/geokrety-website-qa/issues/12:

- [x] setup travis token as environment variable
- [x] add travis trigger build shell
- [x] add travis qa step (todo: to test)
- [x] add on branch condition 😁

travis API to request build: https://docs.travis-ci.com/user/triggering-builds/

--- 

related branch: `feature/webhookQA`

### PR DESCRIPTION
**Pull request by boly38**
Related to #193 chain QA test within travis build

-- setup TRAVIS_QA_TOKEN (on travis side)
- add wait testQA external build 5 minutes
- add notification on irc

---
Please note that the current branch (feature/webhookQA) is not matching qaTest stage conditions. That's why this tests has been done previously. cf #193 

for review purpose, if you were seeing day after day git diff, you could rely on temp branch : `temp_for_review_before_squash_rebase_webhook_qa_v0`

---
please note that this PR could help travis users following [this post](https://stackoverflow.com/questions/32118067/triggering-builds-of-dependent-projects-in-travis-ci)

### CODE REVIEW

**Comment by boly38** [comment]
image

@kumy, could you explain me why there is 2 build for  each PR creation ?
- I understand the `travis-ci/push` one that match the branch commit I would like to merge in master
- but for the first one `travis-ci/pr` seems to be master branch + merge ? this doesn't activated "real deployment" for "master" ? [builds/445064632](https://travis-ci.org/geokrety/geokrety-website/builds/445064632) doesn't update [rec](https://rec.geokrety.org/) ?

**Comment by kumy** [comment]
> why there is 2 builds for each PR creation ?

Because I/we don't use our personal project fork to push to development branches - the push trigger is launched - but this is useful for at least the master branch which upload and download from crowdin automatically.

On another side as we use PR, the other trigger is launched.

**Comment by kumy** [comment]
In build https://travis-ci.org/geokrety/geokrety-website/builds/445064632, the "deploy" stage is used to export/compile translations.

The deployment on rec/boly38/prod is done using webhooks

File: .travis_wait_build.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    echo " * `date '+%Y-%m-%d %H:%M:%S'` [${ITERATION}] requesting ${QA_SLOG} builds";
+    BUILD_INFO=$(curl -s -X GET \
+       -H "Content-Type: application/json" \
+       -H "Accept: application/json" \
+       -H "Travis-API-Version: 3" \
+       -H "Authorization: token ${QA_TOKEN}" \
+       https://api.travis-ci.org/repo/${QA_SLUG}/builds?include=build.state\&include=build.id\&include=build.started_at\&branch.name=master\&sort_by=started_atdesc\&limit=1 )
+
+    if [[ ${BUILD_INFO} == *"access denied"* ]]; then
+      echo " X build #${QA_BUILD_ID} access denied"
```
Comment (boly38) [inline .travis_wait_build.sh]: QA_BUILD_ID not initialized here. remove it

File: .travis_wait_build.sh [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# ToCheck over the time : possible improvement using build_id to ignore instead of waiting intermediate state
+# wait 'created ' or 'started' state
+while [[ ( "${BUILD_STATE}" != "created" ) && ( "${BUILD_STATE}" != "started" ) && ( ${ITERATION} -lt ${ITERATION_MAX} ) ]]
+do
+    if [ ${ITERATION} -ne 0 ]; then
+      sleep ${ITERATION_RETRY_SLEEP}
+    fi
+    getBuildInfo
+    echo " * `date '+%Y-%m-%d %H:%M:%S'` https://travis-ci.org/${QA_SLOG}/builds/${BUILD_ID} => ${BUILD_STATE}"
+    ITERATION=$((ITERATION + 1))
```
Comment (boly38) [inline .travis_wait_build.sh]: dont increment if created or started (to avoid wrong iteration count at the end)

File: .travis_wait_build.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+# wait 'passed' or 'failed' final state
+while [[ ( "${BUILD_STATE}" != "passed" ) && ( "${BUILD_STATE}" != "failed" ) && ( ${ITERATION} -lt ${ITERATION_MAX} ) ]]
+do
+    if [ ${ITERATION} -ne 0 ]; then
+      sleep ${ITERATION_RETRY_SLEEP}
+    fi
+    getBuildInfo
+    echo " * `date '+%Y-%m-%d %H:%M:%S'` https://travis-ci.org/${QA_SLOG}/builds/${BUILD_ID} => ${BUILD_STATE}"
+    ITERATION=$((ITERATION + 1))
```
Comment (boly38) [inline .travis_wait_build.sh]: here too

File: .travis_wait_build.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+ITERATION=0
+
+function getBuildInfo() {
+    echo " * `date '+%Y-%m-%d %H:%M:%S'` [${ITERATION}] requesting ${QA_SLOG} builds";
+    BUILD_INFO=$(curl -s -X GET \
+       -H "Content-Type: application/json" \
+       -H "Accept: application/json" \
+       -H "Travis-API-Version: 3" \
+       -H "Authorization: token ${QA_TOKEN}" \
+       https://api.travis-ci.org/repo/${QA_SLUG}/builds?include=build.state\&include=build.id\&include=build.started_at\&branch.name=master\&sort_by=started_atdesc\&limit=1 )
```
Comment (kumy) [inline .travis_wait_build.sh]: `branch.name=master` shouldn't it be `branch.name=${QA_BRANCH}`?

File: .travis.yml [hunk 5]
```diff
@@ -47,3 +47,20 @@ jobs:
       script:
         - scripts/export-translations-github.sh
         - scripts/compile-translations.sh
+    - stage: qaTests
+      if: branch IN (master, boly38)
```
Comment (kumy) [inline .travis.yml]: Nice idea to use the `if` statement directcly in `.travis.yml`, the same could be applied to `scripts/export-translations-github.sh` and `scripts/compile-translations.sh` scripts.

File: .travis_wait_build.sh (hunk 4 above)
Comment (boly38) [reply .travis_wait_build.sh]: we are talking about geokrety-website-qa branch here

I plant to maybe make it as env variable in the future
but for now we would like to tests using master branch of qa

Comment (kumy) [reply .travis_wait_build.sh]: Ok, I didn't get it :)

**Comment by kumy** [comment]
> why there is 2 builds for each PR creation ?

What about playing with [safelisting-or-blocklisting-branches](https://docs.travis-ci.com/user/customizing-the-build/#safelisting-or-blocklisting-branches) the limit this effect?

**Comment by boly38** [comment]
> > why there is 2 builds for each PR creation ?
> 
> What about playing with [safelisting-or-blocklisting-branches](https://docs.travis-ci.com/user/customizing-the-build/#safelisting-or-blocklisting-branches) the limit this effect?

yep I've found related help in Travis repo settings

but what to  safelist / exclude ? it 's stay not clear for me

for ex. we are interested to keep build for 'feature/xxx' branch push and PR
I dont want to disable these 2 trgiggers
Just don't need these 2 when they're applicable at the same point.

beuh. :)

**Comment by kumy** [comment]
We can start first by whitelisting push build to master/rec/boly38?

**Comment by boly38** [comment]
> We can start first by whitelisting push build to master/rec/boly38?

I think if we do that, travis will not build feature branch :(

Found that section about double build:
https://docs.travis-ci.com/user/pull-requests#double-builds-on-pull-requests

>If you see two build status icons on your GitHub pull request, it means there is one build for the branch, and one build for the pull request itself (actually the build for the merge of the head branch with the base branch specified in the pull request).

I understand that PR build is merge result, which is different from push one (not rebased branch case). So I prefer to keep current state.

**Pull request merged**